provider/resources.go

```diff
@@ -317,65 +322,112 @@ func nameField(info *tfbridge.SchemaInfo) map[string]*tfbridge.SchemaInfo {
 	}
 }
 
-// We should only run the validation once to avoid duplicating the reported errors.
-var credentialsValidationRun atomic.Bool
-
-func preConfigureCallbackWithLogger(ctx context.Context, host *provider.HostClient, vars resource.PropertyMap, c shim.ResourceConfig) error {
-	if !credentialsValidationRun.CompareAndSwap(false, true) {
-		return nil
+func getRegionsList(ctx context.Context, project string, clientOpts []option.ClientOption) ([]string, error) {
+	computeService, err := compute.NewService(ctx, clientOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create compute service: %w", err)
 	}
-	project := tfbridge.ConfigStringValue(vars, "project", []string{
-		"GOOGLE_PROJECT",
-		"GOOGLE_CLOUD_PROJECT",
-		"GCLOUD_PROJECT",
-		"CLOUDSDK_CORE_PROJECT",
-	})
-	if project == "" {
-		msg := "unable to detect a global setting for GCP Project;\n" +
-			"Pulumi will rely on per-resource settings for this operation.\n" +
-			"Set the GCP Project by using:\n" +
-			"\t`pulumi config set gcp:project <project>`"
-		host.Log(ctx, diag.Warning, "", msg) // the URN will default to the root stack name which is exactly what we want
-		//return fmt.Errorf("unable to find required configuration setting: GCP Project\n" +
-		//"Set the GCP Project by using:\n" +
-		//"\t`pulumi config set gcp:project <project>`")
-		//
-		return nil
+	regionsService := compute.NewRegionsService(computeService)
+	regionList, err := regionsService.List(project).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list regions: %w", err)
+	}
+	var regions []string
+	for _, region := range regionList.Items {
+		regions = append(regions, region.Name)
+	}
+	return regions, nil
+}
+
+//go:embed errors/no_credentials.txt
+var noCredentialsErr string
+
+//go:embed errors/wrong_region.txt
+var wrongRegionErr string
+
+//go:embed errors/no_project.txt
+var noProjectErr string
+
+func logOrPrint(ctx context.Context, host *provider.HostClient, msg string) {
+	// host is unavailable in tests, so we revert to normal logging.
+	if host != nil {
+		// the URN will default to the root stack name which is exactly what we want
+		_ = host.Log(ctx, diag.Warning, "", msg)
+	} else {
+		log.Print(msg)
 	}
+}
 
-	config := tpg_transport.Config{
-		AccessToken: tfbridge.ConfigStringValue(vars, "accessToken", []string{"GOOGLE_OAUTH_ACCESS_TOKEN"}),
-		Credentials: tfbridge.ConfigStringValue(vars, "credentials", []string{
-			"GOOGLE_CREDENTIALS",
-			"GOOGLE_CLOUD_KEYFILE_JSON",
-			"GCLOUD_KEYFILE_JSON",
-		}),
-		ImpersonateServiceAccount: tfbridge.ConfigStringValue(vars, "impersonateServiceAccount", []string{"GOOGLE_IMPERSONATE_SERVICE_ACCOUNT"}),
-		Project: tfbridge.ConfigStringValue(vars, "project", []string{
+func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpClientOpts []option.ClientOption) func(
```

Comment:
Is the second argument just for testing?

If so let's make sure there is a doc comment explaining that.